src/hotspot/share/gc/shared/space.inline.hpp

```diff
@@ -175,6 +171,9 @@ inline void CompactibleSpace::scan_and_forward(SpaceType* space, CompactPoint* c
       cur_obj += size;
       end_of_live = cur_obj;
     } else {
+      assert(!space->scanned_block_is_obj(cur_obj) || oop(cur_obj)->mark_raw().is_unlocked() ||
+             oop(cur_obj)->mark_raw().has_bias_pattern() || oop(cur_obj)->mark_raw().has_monitor(),
+             "these are the only valid states during a mark sweep");
```

Comment:
Is this change related? Also, when moving the assert into the else block it will become always-true because of space->scanned_block_is_obj(cur_obj), or am I missing something?

Reply:
See this comment from Coleen and the replies:
https://github.com/openjdk/jdk/pull/135#discussion_r487300636

Please let me know if that resolved this comment for you.

Reply:
Sorry, no. Maybe it's too late here and I shall think about it tomorrow morning instead ;-) Or maybe you can explain it again in the context of that change. How's the assert even relevant when moved in the else-branch?

Reply:
Sorry, I confused myself switching between this review and a
preliminary review thread.

Here's the original code:

165   while (cur_obj < scan_limit) {
166     assert(!space->scanned_block_is_obj(cur_obj) ||
167            oop(cur_obj)->mark_raw().is_marked() || oop(cur_obj)->mark_raw().is_unlocked() ||
168            oop(cur_obj)->mark_raw().has_bias_pattern(),
169            "these are the only valid states during a mark sweep");
170     if (space->scanned_block_is_obj(cur_obj) && oop(cur_obj)->is_gc_marked()) {

and here's the code after it was moved and rewritten:

173     } else {
174       assert(!space->scanned_block_is_obj(cur_obj) || oop(cur_obj)->mark_raw().is_unlocked() ||
175              oop(cur_obj)->mark_raw().has_bias_pattern() || oop(cur_obj)->mark_raw().has_monitor(),
176              "these are the only valid states during a mark sweep");


Where the assert() was worked fine when the ObjectMonitor had a regular oop,
but after it was changed into a weak handle, that location became exposed to
the fact that the object reference could be GC'ed. The original code assumes
that the ObjectMonitor oop reference is stable and unchanging and that's no
longer the case with the weak handle.

Reply:
I found my original preliminary code review comment and Erik's reply:

> src/hotspot/share/gc/shared/space.inline.hpp
>     old L166:     assert(!space->scanned_block_is_obj(cur_obj) ||
>     old L167:            oop(cur_obj)->mark_raw().is_marked() || oop(cur_obj)->mark_raw().is_unlocked() ||
>     old L168: oop(cur_obj)->mark_raw().has_bias_pattern(),
>     old L169:            "these are the only valid states during a mark sweep");
>         This assert was before the if-statement at the top of the while-loop.
>
>     new L174: assert(!space->scanned_block_is_obj(cur_obj) || oop(cur_obj)->mark_raw().is_unlocked() ||
>     new L175: oop(cur_obj)->mark_raw().has_bias_pattern() || oop(cur_obj)->mark_raw().has_monitor(),
>     new L176:              "these are the only valid states during a mark sweep");
>         The assert is now in the else branch of the following if-statement:
>
>            L166     if (space->scanned_block_is_obj(cur_obj) && oop(cur_obj)->is_gc_marked()) {
>
>         The new assert() drops this check:
>
>             oop(cur_obj)->mark_raw().is_marked()
>
>         and adds this check:
>
>             oop(cur_obj)->mark_raw().has_monitor()
>
>         Dropping the "is_marked()" makes sense since the new location
>         of the assert is in the else branch of "oop(cur_obj)->is_gc_marked()".
>         The addition of the "has_monitor()" check is puzzling. Why was this
>         added and why wasn't it needed in the old assert()?
>
>         In fact, I'm not sure why this change is here at all.

This is an artifact of the monitor now being weak. Since there was previously always a strong root
to all inflated monitors, there were never any dead objects in the heap, that still had pointers in the
mark word to the monitor. The change to weak now implies that we suddenly have dead objects
in the heap, that in the markWord point out their monitor. GC code that iterates through consecutive
objects one by one, will see these now dead objects with monitors. The assert changes reflect that.
Before it was unexpected and would assert on that. Now I moved the assertion to the case when the
object is alive instead. We have no business asserting what should be in the markWord of dead objects.

I hope it makes sense now!

Reply:
This code seems like something that doesn't belong here anymore.  This code assumed synchronous scanning of oops in ObjectMonitor and scanning memory regions, and that's no longer the case with OopStorage. I think this assert should be removed.  It exports some implementation detail of now completely unrelated code in order to do a very specific check.

Reply:
@fisk - please chime in here...

Reply:
I agree. Also, the assert becomes true somewhat obviously because of its first clause, which is already guaranteed because of its placement in the surrounding else-branch (unless something really weird happens).

Reply:
I am also 100% okay with just removing the random assertion. We have a whole bunch of other block iteration code in GC code that does not sprinkle random asserts about what patterns of markWords are supposedly good or bad in live/dead objects. And I also don't get what kind of bug this would catch.

Reply:
Okay. Since @coleenp, @rkennke and @fisk want that assertion to "hit the road",
I'm in the process of removing it... build running now (just 'cause I'm paranoid me...)...

Reply:
This sub-thread should be resolve via https://github.com/openjdk/jdk/pull/135/commits/eeb9d761374ea193a7295fcce3ffe9707a8f0348